testbed/tests/app/test_desktop.py

```diff
@@ -370,26 +368,47 @@ async def test_current_window(app, app_probe, main_window):
     window2.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
     window3.content = toga.Box(style=Pack(background_color=FIREBRICK))
 
-    # We don't need to probe anything window specific; we just need
-    # a window probe to enforce appropriate delays.
-    window1_probe = window_probe(app, window1)
-
     window1.show()
     window2.show()
     window3.show()
 
-    await window1_probe.wait_for_window("Extra windows added")
+    await main_window_probe.wait_for_window("Extra windows added")
 
+    # When a window without any dialog is made the current_window,
+    # then `app.current_window` should return the specified window.
     app.current_window = window2
-    await window1_probe.wait_for_window("Window 2 is current")
+    await main_window_probe.wait_for_window("Window 2 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window2
 
     app.current_window = window3
-    await window1_probe.wait_for_window("Window 3 is current")
+    await main_window_probe.wait_for_window("Window 3 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window3
 
+    # When a dialog is in focus, `app.current_window` should
+    # return the window from which the dialog was initiated.
+    info_dialog = toga.InfoDialog("Info", "Some info")
+    app_probe.setup_info_dialog_result(info_dialog)
+
+    app.current_window = window1
+    await main_window_probe.wait_for_window("Window 1 is current")
+    dialog_task = app.loop.create_task(window1.dialog(info_dialog))
+    await main_window_probe.wait_for_window("Displayed window1 modal info dialog")
```

Comment:
I'm not convinced this is a good test. 

Firstly, the assignment to window1 is redundant - there's nothing special about window 1, so that's a step that isn't required.

But, more importantly, this entirely depends on the exact sequencing of events. It's using `create_task()` to schedule a coroutine on the event loop - and then *hoping* that all the checks about window assignment are performed before that task completes. 

In practice, that seems highly likely... but it's not guaranteed. And it also leads to the "dangling task" issue, as there's no guarantee that the task will have ended by the time the test is cleaning up.

A better approach would be to modify the `setup_info_dialog_result` probe to accept an optional method that contains "pre-close" test logic, and invoke that method as required before closing.

Reply:
Apologies for the delayed response; I got caught up with a few things. Anyway, Thank you for suggesting the test strategy.